lte/gateway/python/integ_tests/s1aptests/s1ap_utils.py

```diff
@@ -1049,45 +1049,27 @@ def restart_services(self, services, wait_time=0):
             services: List of (str) services names
             wait_time: (int) Time to wait for restart of the services
         """
-        for s in services:
-            if s == "mme":
-                self.restart_mme(0)
-            elif s == "sctpd":
-                self.restart_sctpd(0)
-            else:
-                if self._init_system == InitMode.SYSTEMD:
-                    self.exec_command(f"sudo systemctl restart magma@{s}")
-                elif self._init_system == InitMode.DOCKER:
-                    self.exec_command(f"docker restart {s}")
-        self.wait_for_restart_to_finish(wait_time)
+        for service in services:
+            service_name = self.get_service_name_from_init_system(service)
+            if self._init_system == InitMode.SYSTEMD:
+                self.exec_command(f"sudo systemctl restart {service_name}")
+            elif self._init_system == InitMode.DOCKER:
+                if service_name == "oai_mme":
+                    self.exec_command(
+                        f"docker restart mobilityd pipelined sessiond oai_mme",
+                    )
+                elif service_name == "sctpd":
```

Comment:
Does this restart mobilityd, pipelined, sessiond and oai_mme twice if they are together with sctpd in the services list? is this intentional? Same question for oai_mme above.

Reply:
Yes. We didn't spent any time to polish this so it is more of an hacky solution that works than really intentional.

Reply:
Good catch. For now none of the testcases actually call restart of mme and sctpd together. SCTPD is restarted only if we need explicitely sctpd restart or MME restart with fluching redis data as well because the situation is known that sctpd restart will internally restart MME also. However we have 2 testcases where multiple services along with mme are restarted. Definitely need to check on them if that is taken care not to be duplicate here:
```
s1aptests/test_attach_ul_udp_data_with_multiple_service_restart.py
  > ["mobilityd", "mme", "pipelined"]
s1aptests/test_secondary_pdn_with_dedicated_bearer_multiple_services_restart.py
  > ["sessiond", "mme", "pipelined"]
```

Reply:
At the moment we explicitly restart the `mobilityd`, `pipelined`, and `sessiond` containers if we restart the `oai-mme`, which leads to a double-restart in these tests. But I don't think that this is different compared to what's happening in the `systemd` case. Only now, it is directly stated in the restart command instead of hidden somewhere in the service dependencies in the `.yaml` config files.
```
vagrant@magma-dev:~/magma/lte/gateway/docker$ sudo systemctl show magma@mobilityd magma@pipelined magma@sessiond --property ActiveEnterTimestamp
ActiveEnterTimestamp=Thu 2022-09-29 14:22:03 UTC

ActiveEnterTimestamp=Thu 2022-09-29 14:22:03 UTC

ActiveEnterTimestamp=Thu 2022-09-29 14:22:02 UTC
vagrant@magma-dev:~/magma/lte/gateway/docker$ sudo systemctl restart magma@mme
vagrant@magma-dev:~/magma/lte/gateway/docker$ sudo systemctl show magma@mobilityd magma@pipelined magma@sessiond --property ActiveEnterTimestamp
ActiveEnterTimestamp=Thu 2022-09-29 14:24:17 UTC

ActiveEnterTimestamp=Thu 2022-09-29 14:24:17 UTC

ActiveEnterTimestamp=Thu 2022-09-29 14:24:16 UTC
```
The way I see this, these tests always have restarted `mme` plus some other services which `mme` restarts anyway. Or am I missing something here?

Reply:
I agree - `magma@mme.service` has
```
PartOf=magma@pipelined.service
PartOf=magma@sessiond.service
```
this causes a restart of `pipelined` and `sessiond` if `mme` is restarted.

`magma@pipelined.service` and `magma@sessiond.service` also have vice versa `PartOf=magma@mme.service`.

@mpfirrmann Is there a similar logic in the dockerized agw for the container?

Imho in `test_attach_ul_udp_data_with_multiple_service_restart` and `test_secondary_pdn_with_dedicated_bearer_multiple_services_restart` the explicit restart of `sessiond` and `pipelined` should be removed and a comment should be added that the restart of `mme` causes a cascade of restarts (actually a restart of `mme` in the systemd case causes a restart of `sessiond`, `pipelined`, `mobilityd`, `liagent`, `envoy` and `envoy-controller` (not all are running in vm setup)). See also https://excalidraw.com/#room=9b3ea34b6f337ccab568,F2DMzJxlW8yjKZtjQ2djtA
@VinashakAnkitAman what do you think? (I see this out of scope here tho)

Detailed systemd logging of `for s in magma@mme magma@sessiond; do sudo systemctl restart $s; done` - we see that mme is restarted twice:
```
Sep 29 16:51:28 magma-deb systemd[1]: Stopping Magma OAI MME service...
Sep 29 16:51:28 magma-deb systemd[1]: Stopping Magma envoy service...
Sep 29 16:51:29 magma-deb systemd[1]: magma@mme.service: Main process exited, code=exited, status=23/n/a
Sep 29 16:51:29 magma-deb systemd[1]: magma@mme.service: Failed with result 'exit-code'.
Sep 29 16:51:29 magma-deb systemd[1]: Stopped Magma OAI MME service.
Sep 29 16:51:29 magma-deb systemd[1]: Stopping Magma pipelined service...
Sep 29 16:51:29 magma-deb systemd[1]: Stopping Magma session manager service...
Sep 29 16:51:30 magma-deb systemd[1]: magma@sessiond.service: Succeeded.
Sep 29 16:51:30 magma-deb systemd[1]: Stopped Magma session manager service.
Sep 29 16:51:30 magma-deb systemd[1]: Started Magma session manager service.
Sep 29 16:51:30 magma-deb systemd[1]: magma@pipelined.service: Succeeded.
Sep 29 16:51:30 magma-deb systemd[1]: Stopped Magma pipelined service.
Sep 29 16:51:30 magma-deb systemd[1]: Stopping Magma mobilityd service...
Sep 29 16:51:30 magma-deb systemd[5392]: run-netns-envoy_ns1.mount: Succeeded.
Sep 29 16:51:30 magma-deb systemd[1]: run-netns-envoy_ns1.mount: Succeeded.
Sep 29 16:51:30 magma-deb systemd[1]: magma_dp@envoy.service: Succeeded.
Sep 29 16:51:30 magma-deb systemd[1]: Stopped Magma envoy service.
Sep 29 16:51:30 magma-deb systemd[1]: Stopping Magma envoy controller service...
Sep 29 16:51:30 magma-deb systemd[1]: magma@mobilityd.service: Succeeded.
Sep 29 16:51:30 magma-deb systemd[1]: Stopped Magma mobilityd service.
Sep 29 16:51:30 magma-deb systemd[1]: Started Magma mobilityd service.
Sep 29 16:51:30 magma-deb systemd[1]: magma@envoy_controller.service: Succeeded.
Sep 29 16:51:30 magma-deb systemd[1]: Stopped Magma envoy controller service.
Sep 29 16:51:30 magma-deb systemd[1]: Started Magma envoy controller service.
Sep 29 16:51:30 magma-deb systemd[1]: Starting Magma pipelined service...
Sep 29 16:51:30 magma-deb systemd[1]: Starting Magma envoy service...
Sep 29 16:51:30 magma-deb systemd[1]: Started Magma pipelined service.
Sep 29 16:51:30 magma-deb systemd[1]: Starting Magma OAI MME service...
Sep 29 16:51:31 magma-deb systemd[1]: Started Magma OAI MME service.
Sep 29 16:51:31 magma-deb systemd[1]: Stopping Magma OAI MME service...
Sep 29 16:51:31 magma-deb systemd[1]: magma@mme.service: Succeeded.
Sep 29 16:51:31 magma-deb systemd[1]: Stopped Magma OAI MME service.
Sep 29 16:51:31 magma-deb systemd[1]: Stopping Magma pipelined service...
Sep 29 16:51:31 magma-deb systemd[1]: Stopping Magma session manager service...
Sep 29 16:51:32 magma-deb systemd[1]: magma@pipelined.service: Succeeded.
Sep 29 16:51:32 magma-deb systemd[1]: Stopped Magma pipelined service.
Sep 29 16:51:32 magma-deb systemd[1]: Stopping Magma envoy controller service...
Sep 29 16:51:32 magma-deb systemd[1]: Stopping Magma mobilityd service...
Sep 29 16:51:32 magma-deb systemd[1]: magma@sessiond.service: Succeeded.
Sep 29 16:51:32 magma-deb systemd[1]: Stopped Magma session manager service.
Sep 29 16:51:32 magma-deb systemd[1]: Started Magma session manager service.
Sep 29 16:51:32 magma-deb systemd[1]: magma@envoy_controller.service: Succeeded.
Sep 29 16:51:32 magma-deb systemd[1]: Stopped Magma envoy controller service.
Sep 29 16:51:32 magma-deb systemd[1]: Started Magma envoy controller service.
Sep 29 16:51:32 magma-deb systemd[1]: magma@mobilityd.service: Succeeded.
Sep 29 16:51:32 magma-deb systemd[1]: Stopped Magma mobilityd service.
Sep 29 16:51:32 magma-deb systemd[1]: Started Magma mobilityd service.
Sep 29 16:51:32 magma-deb systemd[1]: Starting Magma pipelined service...
Sep 29 16:51:32 magma-deb systemd[1]: Started Magma pipelined service.
Sep 29 16:51:32 magma-deb systemd[1]: Starting Magma OAI MME service...
Sep 29 16:51:33 magma-deb systemd[1]: Started Magma OAI MME service.
```

Reply:
I adapted the restarts of the docker container according to the `systemd` dependencies. This is for restarting and stopping services/containers. Starting a single `systemd`-magma service does not start the others. 
The restarts in the two mentioned tests was also adapted.

While this makes both init system equivalent in testing, the dependencies are missing in production (#14055).
@VinashakAnkitAman, what is your opinion on this?

Reply:
Thanks for marking the Github issues here for tracking linked references and also for adding the linked description in comments in the PR change. The missing dependencies can definitely be taken care under a different PR followed under the above mentioned GH issue

The changes for only mme restart in the mentioned 2 testcases also looks good to me.